lib/core/utils/query-selector-all-filter.js

```diff
@@ -26,13 +26,15 @@ function matchExpressions(domTree, expressions, filter) {
     const vNode = currentLevel.vNodes.pop();
     const childOnly = []; // we will add hierarchical '>' selectors here
     const childAny = [];
-    const combined = currentLevel.anyLevel
-      .slice()
-      .concat(currentLevel.thisLevel);
+    const combinedLength =
+      currentLevel.anyLevel.length + currentLevel.thisLevel.length;
     let added = false;
     // see if node matches
-    for (let i = 0; i < combined.length; i++) {
-      const exp = combined[i];
+    for (let i = 0; i < combinedLength; i++) {
+      const exp =
+        i < currentLevel.anyLevel.length
+          ? currentLevel.anyLevel[i]
+          : currentLevel.thisLevel[i - currentLevel.thisLevel.length];
```

Comment:
This should have also failed a test, but seems we are missing some tests cases if it didn't.

```suggestion
          : currentLevel.thisLevel[i - currentLevel.anyLevel.length];
```

Reply:
Done! For all the existing tests `currentLevel.anyLevel.length === currentLevel.thisLevel.length`, added one where they are not equal.